src/compiler/program.ts

```diff
@@ -2049,6 +2119,69 @@ namespace ts {
             return allFilesBelongToPath;
         }
 
+        function createProjectReferenceRedirects(rootOptions: CompilerOptions): Map<string> {
+            const result = createMap<string>();
+            const handledProjects = createMap<true>();
+
+            walkProjectReferenceGraph(host, rootOptions, createMapping);
+
+            function createMapping(resolvedFile: string, referencedProject: CompilerOptions, reference: ProjectReference) {
+                const rootDir = normalizePath(referencedProject.rootDir || getDirectoryPath(resolvedFile));
+                result.set(rootDir, referencedProject.outDir);

```

Comment:
`outDir` can be empty. do we want to add a mapping to `undefined`?

Reply:
✔️